File: src/drive/files_api.rs
//! Files API —— GET 列举/搜索、POST 创建、PATCH 更新/软删除。
//!
//! 对齐 `legacy/lib/drive/api/files_api.dart`。
//!
//! # 华为 API 怪癖
//! - **parentFolder 查询语法**：不用 parentFolder 参数，而用 `queryParam='root' in parentFolder`
//!   （单引号包裹 token）。列出根目录用 `'root'`，列出子目录用 `'<id>'`。
//! - **asciiJsonEncode**：createFolder / update 的 application/json 请求体必须用 ASCII-only 编码，
//!   否则中文名报 400 `21004002 fileName can not be blank`。
//! - **createFolder**：mimeType 必填，root 目录省略 parentFolder。

/// 文件读取与严格分页实现。
mod read;
/// 文件请求参数校验与编码。
mod request;
/// 文件响应协议解析与写后核验。
mod response;
/// 文件创建、更新和软删除实现。
mod write;

use std::sync::Arc;

use crate::error::{AppError, AppResult};

pub use request::{build_create_folder_body, urlencoding};

/// 生产环境目录全量分页的最大请求页数。
const PRODUCTION_MAX_PAGES: usize = 1_000;

/// Files:list 的客户端分页上限。
///
/// 华为只定义单页大小上限，没有定义目录总页数。客户端仍需要有限上限来避免服务端
/// cursor 循环或异常数据导致永久索引；达到上限且仍有下一页时必须失败，不能返回部分树。
#[derive(Debug, Clone, Copy)]
pub struct PaginationPolicy {
    max_pages: usize,
}

impl PaginationPolicy {
    /// 创建非零分页上限，否则返回配置错误。
    pub fn new(max_pages: usize) -> AppResult<Self> {
        if max_pages == 0 {
            return Err(AppError::generic("Files 分页上限必须大于 0"));
        }
        Ok(Self { max_pages })
    }

    /// 返回生产环境分页策略。
    const fn production() -> Self {
        Self {
            max_pages: PRODUCTION_MAX_PAGES,
        }
    }
}

/// 提供严格分页、解析及写后核验的云盘文件接口。
pub struct FilesApi {
    client: Arc<crate::drive::client::DriveClient>,
    pagination: PaginationPolicy,
}

impl FilesApi {
    /// 使用生产分页上限创建文件接口。
    pub fn new(client: Arc<crate::drive::client::DriveClient>) -> Self {
        Self {
            client,
            pagination: PaginationPolicy::production(),
        }
    }

    /// 使用可控的分页上限构造真实 Files API wrapper。
    ///
    /// 该 seam 仍走 [`DriveClient`] 的生产请求链，只替换防无限分页的客户端上限。
    pub fn with_pagination_policy(
        client: Arc<crate::drive::client::DriveClient>,
        pagination: PaginationPolicy,
    ) -> Self {
        Self { client, pagination }
    }
}

File: shared/src/commonMain/kotlin/io/github/yuanbaobaao/petallink/drive/FilesApi.kt
package io.github.yuanbaobaao.petallink.drive

import io.github.yuanbaobaao.petallink.auth.Pkce
import io.ktor.client.request.header
import io.ktor.client.request.setBody
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

/**
 * 华为 Files API（对标 src/drive/files_api.rs）。
 *
 * 详见 docs/03 §Files API、docs/10 阶段 2 item 9。
 * 踩坑：parentFolder 用 queryParam 语法（'id' in parentFolder），非参数。
 */
class FilesApi(
    private val client: DriveClient,
    private val base: String = DriveConstants.DRIVE_API_BASE,
) {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    /** 列出文件（分页） */
    suspend fun listFiles(
        parentId: String?,
        pageSize: Int = 100,
        nextCursor: String? = null,
    ): Pair<List<DriveFile>, String?> {
        val params = mutableListOf("pageSize" to pageSize.toString())
        // 踩坑 5：List 用 queryParam='id' in parentFolder，非 parentFolder 参数
        if (parentId != null) {
            params += "queryParam" to "'id' in parentFolder"
        }
        nextCursor?.let { params += "cursor" to it }
        val qs = params.joinToString("&") { "${it.first}=${it.second}" }

        val resp = client.executeWithRetry(
            HttpMethod.Get, "$base/files?$qs", HttpSemantics.READ,
        )
        val body = Json.parseToJsonElement(resp.bodyAsText()).jsonObject
        val files = body["files"]?.jsonArray?.map { parseDriveFile(it.jsonObject) } ?: emptyList()
        val cursor = body["nextCursor"]?.jsonPrimitive?.contentOrNull
        return files to cursor
    }

    /** 获取单个文件（fields=* 全字段） */
    suspend fun getFile(fileId: String): DriveFile {
        val resp = client.executeWithRetry(
            HttpMethod.Get,
            "$base/files/${Pkce.enc(fileId)}?fields=*",
            HttpSemantics.READ,
        )
        val body = Json.parseToJsonElement(resp.bodyAsText()).jsonObject
        return parseDriveFile(body)
    }

    /** 创建文件/文件夹 */
    suspend fun createFile(name: String, parentFolder: String?, isFolder: Boolean): DriveFile {
        val meta = buildJsonObject {
            // 踩坑 4：中文文件名在 application/json 需 ascii 转义
            put("fileName", AsciiJson.escape(name))
            if (parentFolder != null) putJsonArray("parentFolder") { add(parentFolder) }
            if (isFolder) put("mimeType", "application/vnd.huawei-apps.folder")
        }
        val resp = client.executeWithRetry(
            HttpMethod.Post, "$base/files", HttpSemantics.WRITE,
        ) {
            contentType(ContentType.Application.Json)
            setBody(meta.toString())
        }
        // 踩坑 8：写操作必须 200（require_official_write_ok）
        require(resp.status.value == 200) {
            throw io.github.yuanbaobaao.petallink.AppError.Remote(
                resp.status.value, "create 未返回 200"
            )
        }
        return parseDriveFile(Json.parseToJsonElement(resp.bodyAsText()).jsonObject)
    }

    /** 更新文件（PATCH 覆盖） */
    suspend fun updateFile(fileId: String, name: String?): DriveFile {
        val meta = buildJsonObject {
            if (name != null) put("fileName", AsciiJson.escape(name))
        }
        val resp = client.executeWithRetry(
            HttpMethod.Patch, "$base/files/${Pkce.enc(fileId)}", HttpSemantics.WRITE,
        ) {
            contentType(ContentType.Application.Json)
            setBody(meta.toString())
        }
        require(resp.status.value == 200) {
            throw io.github.yuanbaobaao.petallink.AppError.Remote(
                resp.status.value, "update 未返回 200"
            )
        }
        return parseDriveFile(Json.parseToJsonElement(resp.bodyAsText()).jsonObject)
    }

    /** 软删除（PATCH recycled=true，踩坑：非物理删除） */
    suspend fun deleteFile(fileId: String) {
        val body = buildJsonObject { put("recycled", true) }
        val resp = client.executeWithRetry(
            HttpMethod.Patch, "$base/files/${Pkce.enc(fileId)}", HttpSemantics.WRITE,
        ) {
            contentType(ContentType.Application.Json)
            setBody(body.toString())
        }
        // 软删除成功合同：200 + File(id==请求 id, recycled==true)
        require(resp.status.value == 200) {
            throw io.github.yuanbaobaao.petallink.AppError.Remote(
                resp.status.value, "delete 未返回 200"
            )
        }
    }

    /**
     * 列出全部文件（自动翻页，1000 页上限）。
     */
    suspend fun listAllFiles(parentId: String?, maxPages: Int = 1000): List<DriveFile> {
        val all = mutableListOf<DriveFile>()
        var cursor: String? = null
        repeat(maxPages) {
            val (files, next) = listFiles(parentId, nextCursor = cursor)
            all += files
            cursor = next
            if (cursor == null) return all
        }
        throw io.github.yuanbaobaao.petallink.AppError.Remote(
            0, "listAllFiles 达到页数上限 $maxPages"
        )
    }

    /** 解析 DriveFile（容忍 String 类型的 size，踩坑 6） */
    private fun parseDriveFile(obj: JsonObject): DriveFile {
        return json.decodeFromJsonElement(DriveFile.serializer(), obj)
    }
}

